assets/src/auth/AuthContext.tsx

```diff
@@ -1,49 +1,154 @@
-import * as React from "react";
+import React, { useState, useEffect, useContext } from "react";
+import createAuth0Client from "@auth0/auth0-spa-js";
+import Auth0Client from "@auth0/auth0-spa-js/dist/typings/Auth0Client";
 
-interface State {
-  email: string;
-  session: string;
+import "../apollo-client";
+
+export interface Auth0RedirectState {
+  targetUrl?: string;
 }
 
-export interface AuthContextValues extends State {
-  setEmail(email: string): void;
-  setSession(session: string): void;
+export interface Auth0User extends Omit<IdToken, "__raw"> {}
+
+interface Auth0Context {
+  user?: Auth0User;
+  isAuthenticated: boolean;
+  isInitializing: boolean;
+  isPopupOpen: boolean;
+  loginWithPopup(o?: PopupLoginOptions): Promise<void>;
+  handleRedirectCallback(): Promise<RedirectLoginResult>;
+  getIdTokenClaims(o?: getIdTokenClaimsOptions): Promise<IdToken>;
+  loginWithRedirect(o?: RedirectLoginOptions): Promise<void>;
+  getTokenSilently(o?: GetTokenSilentlyOptions): Promise<string | undefined>;
+  getTokenWithPopup(o?: GetTokenWithPopupOptions): Promise<string | undefined>;
+  logout(o?: LogoutOptions): void;
 }
 
-const initialState: State = {
-  email: "",
-  session: ""
-};
+// these props duplicate Auth0ClientOptions
+//
+// we only enumerate here them because useEffect is complicated and
+// not worth the effort of getting right
+interface Auth0ProviderOptions {
+  onRedirectCallback(result: RedirectLoginResult): void;
+  children: React.ReactElement;
+  domain: string;
+  client_id: string;
+  audience?: string;
+  redirect_uri?: string;
+}
 
-export const authContextDefaults: AuthContextValues = {
-  ...initialState,
-  setEmail: (email: string) => {},
-  setSession: (session: string) => {}
-};
+export const Auth0Context = React.createContext<Auth0Context | null>(null);
+export const useAuth0 = () => useContext(Auth0Context)!;
+export const Auth0Provider = ({
+  children,
+  onRedirectCallback,
+  client_id,
+  audience,
+  domain,
+  redirect_uri
+}: Auth0ProviderOptions) => {
+  const [isAuthenticated, setIsAuthenticated] = useState(false);
+  const [isInitializing, setIsInitializing] = useState(true);
+  const [isPopupOpen, setIsPopupOpen] = useState(false);
+  const [user, setUser] = useState<Auth0User>();
+  const [auth0Client, setAuth0Client] = useState<Auth0Client>();
+
+  useEffect(() => {
+    const initAuth0 = async () => {
+      const auth0FromHook = await createAuth0Client({
+        client_id,
+        audience,
+        domain,
+        redirect_uri
+      });
+      setAuth0Client(auth0FromHook);
+
+      if (window.location.search.includes("code=")) {
+        let appState: RedirectLoginResult = {};
+        try {
+          ({ appState } = await auth0FromHook.handleRedirectCallback());
+        } finally {
+          onRedirectCallback(appState);
```

Comment:
is it intentional that `onRedirectCallback(appState)` is always run even on error? also, do you need to put a catch in here to handle any errors or is it just bubbling upwards if it throws?

Reply:
not sure. got this from a github issue; haven't gone through and massaged the code yet.

https://github.com/auth0/auth0-spa-js/issues/39
https://github.com/sferguson12/auth0-typescript-react